File: sdk-java/src/main/java/com/flux/Validation.java
/*
 * Validacion L3 contra el JSON Schema del evento.
 * Contrato normativo: specification/00-protocol.md §5 (nivel L3).
 *
 * Cierra el hueco mas grande que quedaba en L2: sin esto, un productor puede publicar un
 * payload que viola su propio `dataschema` y nadie se entera hasta que un consumidor
 * —posiblemente en otro equipo, otro lenguaje y otra semana— se atraganta. El error
 * aparece lejisimos de su causa.
 *
 * Validar en `publish()` lo convierte en un fallo del servicio que lo provoco.
 *
 * ⚠️ La libreria de validacion es una dependencia OPCIONAL del pom
 * (`com.networknt:json-schema-validator`, marcada <optional>true</optional>): L3 es opt-in,
 * asi que su coste tambien debe serlo. Un servicio en L2 no deberia arrastrar un validador
 * de JSON Schema que no va a ejecutar nunca. Por eso todo lo que la toca vive en
 * NetworkntValidator, una clase que solo se CARGA cuando el modo no es OFF: si el jar no
 * esta, el fallo es un mensaje accionable en connect() y no un NoClassDefFoundError a
 * media publicacion.
 */
package com.flux;

/**
 * Modo, bundle y fabrica de la validacion L3.
 *
 * <p>Clase de utilidad: no se instancia.
 *
 * <pre>{@code
 * SchemaBundle bundle = SchemaBundle.fromPath(Path.of("schemas/bundle.json"));
 *
 * FluxBus bus = FluxBus.connect(new FluxBus.ConnectOptions()
 *         // …
 *         .validation(new Validation.Options()
 *                 .mode(Validation.Mode.STRICT)
 *                 .bundle(bundle)
 *                 .onConsume(true)));
 * }</pre>
 */
public final class Validation {

    private Validation() {
        throw new AssertionError("clase de utilidad");
    }

    // ─── Modo ────────────────────────────────────────────────────────────────

    /** Que hacer cuando el payload no cumple su esquema — 00-protocol.md §5. */
    public enum Mode {
        /**
         * Default: nivel L2. No se compila nada y no se paga nada.
         */
        OFF,
        /**
         * Se registra y se publica igual.
         *
         * <p>Existe por la misma razon que el {@code warn} de la firma: introducir
         * validacion en un ecosistema en marcha exige un periodo en el que se ve el
         * incumplimiento sin romper a nadie el primer dia.
         */
        WARN,
        /**
         * {@code publish()} LANZA si el payload no valida. Es el nivel L3 de verdad: un
         * contrato roto pasa a ser un fallo del productor.
         */
        STRICT
    }

    // ─── Opciones ────────────────────────────────────────────────────────────

    /** Configuracion de la validacion L3. */
    public static final class Options {
        private Mode mode = Mode.OFF;
        private SchemaBundle bundle;
        private boolean onConsume;
        private System.Logger logger;

        /** Default {@link Mode#OFF}. */
        public Options mode(Mode v) {
            this.mode = v != null ? v : Mode.OFF;
            return this;
        }

        /**
         * El bundle generado por {@code scripts/bundle-schemas.mjs}.
         *
         * <p>Se pasa como DATO: el SDK no resuelve el {@code dataschema} por HTTP ni con
         * el modo estricto puesto — 00-protocol.md §5, "Resolucion de esquemas: bundle, no
         * HTTP".
         */
        public Options bundle(SchemaBundle v) {
            this.bundle = v;
            return this;
        }

        /**
         * Validar tambien al CONSUMIR. Default {@code false}.
         *
         * <p>Un fallo se clasifica PERMANENT: el evento es sintacticamente correcto pero
         * incumple su contrato, y reintentarlo dara exactamente el mismo resultado
         * — 04-errors.md §1.2.
         */
        public Options onConsume(boolean v) {
            this.onConsume = v;
            return this;
        }

        /**
         * Donde se registran los avisos de {@link Mode#WARN}.
         *
         * <p>{@code null} usa {@code System.getLogger("flux")}, no el silencio: un modo que
         * se llama {@code warn} y no avisa de nada es indistinguible de {@link Mode#OFF}, y
         * quien lo configura cree que esta viendo los incumplimientos.
         */
        public Options logger(System.Logger v) {
            this.logger = v;
            return this;
        }

        /** El modo configurado. */
        public Mode mode() {
            return mode;
        }

        /** El bundle configurado, o {@code null}. */
        public SchemaBundle bundle() {
            return bundle;
        }

        /** Si se valida tambien al consumir. */
        public boolean onConsume() {
            return onConsume;
        }

        System.Logger logger() {
            return logger;
        }
    }

    // ─── Validador ───────────────────────────────────────────────────────────

    /** Comprueba un evento contra el esquema que su {@code dataschema} declara. */
    @FunctionalInterface
    public interface Validator {
        /**
         * Valida el {@code data} del evento.
         *
         * @param event   el evento ya construido (al publicar) o ya parseado (al consumir).
         * @param subject subject del evento, solo para el mensaje de error.
         * @throws SchemaValidationException en {@link Mode#STRICT}, si el payload no cumple.
         * @throws SchemaNotFoundException   en {@link Mode#STRICT}, si el bundle no trae su
         *                                   esquema.
         */
        void check(FluxEvent event, String subject);
    }

    /**
     * Compila los validadores del bundle UNA vez.
     *
     * <p>Lo llama {@code FluxBus.connect()} y no la ruta caliente: compilar un JSON Schema
     * por evento tiraria el throughput y no aportaria nada, porque el bundle es inmutable
     * durante la vida del proceso.
     *
     * @return {@code null} en {@link Mode#OFF}. Un {@code null} aqui es la forma de que L2
     *         no pague absolutamente nada por L3.
     * @throws IllegalArgumentException si el modo no es OFF y falta el bundle.
     * @throws IllegalStateException    si falta la libreria de validacion en el classpath.
     */
    public static Validator create(Options options) {
        Options o = options != null ? options : new Options();
        if (o.mode() == Mode.OFF) {
            return null;
        }
        if (o.bundle() == null) {
            throw new IllegalArgumentException(
                    "flux: el modo de validacion " + o.mode() + " exige un bundle. Generalo con "
                            + "`node scripts/bundle-schemas.mjs` y pasalo con "
                            + "`new Validation.Options().bundle(SchemaBundle.fromPath(...))`");
        }

        System.Logger logger = o.logger() != null ? o.logger() : System.getLogger("flux");
        try {
            // La clase se carga AQUI, en el arranque, y solo si hace falta. Es el
            // equivalente del `await import("ajv/dist/2020.js")` del SDK de Node: la
            // dependencia es opcional, asi que su ausencia tiene que dar un mensaje que
            // diga que instalar — y no un NoClassDefFoundError en la primera publicacion.
            return new NetworkntValidator(o.mode(), o.bundle(), logger);
        } catch (NoClassDefFoundError | ExceptionInInitializerError e) {
            throw new IllegalStateException(
                    "flux: el modo de validacion " + o.mode() + " necesita "
                            + "com.networknt:json-schema-validator en el classpath. Es una "
                            + "dependencia OPCIONAL del SDK porque L3 es opt-in y su coste tambien:\n"
                            + "  <dependency>\n"
                            + "    <groupId>com.networknt</groupId>\n"
                            + "    <artifactId>json-schema-validator</artifactId>\n"
                            + "    <version>" + NetworkntValidator.LIBRARY_VERSION + "</version>\n"
                            + "  </dependency>\n"
                            + "Causa: " + e, e);
        }
    }
}

File: sdk-dotnet/src/Flux/Validation.cs
// Validación L3: contrato y política. El compilador de JSON Schema vive en el paquete
// Flux.Validation.
// Contrato normativo: specification/00-protocol.md §5 (nivel L3).
//
// L3 cierra el hueco más grande que quedaba en L2: sin esto, un productor puede publicar un
// payload que viola su propio `dataschema` y nadie se entera hasta que un consumidor
// —posiblemente en otro equipo, otro lenguaje y otra semana— se atraganta. El error aparece
// lejísimos de su causa. Validar en `PublishAsync()` lo convierte en un fallo del servicio
// que lo provocó.
//
// ⚠️ `System.Text.Json` NO valida JSON Schema. Trae `JsonDocument`, `JsonNode` y
// serialización, pero no un evaluador de esquemas: hace falta un paquete. Es la misma
// situación que con Ed25519 (ver Signing.cs), y se resuelve igual — este fichero contiene
// TODO menos el evaluador: el modo, las excepciones y la interfaz. El paquete
// `Flux.Validation` —que sí arrastra JsonSchema.Net y sus tres dependencias transitivas—
// implementa la interfaz. Así el paquete `Flux`, que es el que instala todo servicio que
// solo quiere publicar un evento, sigue dependiendo únicamente de NATS.Net y
// System.Text.Json. Ver sdk-dotnet/README.md §"Validación L3".

namespace Flux;

/// <summary>Qué hacer cuando el payload no cumple su esquema — 00-protocol.md §5.</summary>
public enum ValidationMode
{
    /// <summary>Default: nivel L2. No se compila nada y no se paga nada.</summary>
    Off,

    /// <summary>
    /// Se registra y se publica igual.
    /// </summary>
    /// <remarks>
    /// Existe por la misma razón que <see cref="VerificationMode.Warn"/>: introducir
    /// validación en un ecosistema en marcha exige un periodo en el que se ve el
    /// incumplimiento sin romper a nadie el primer día.
    /// </remarks>
    Warn,

    /// <summary>
    /// <see cref="FluxBus.PublishAsync"/> LANZA si el payload no valida. Es el nivel L3 de
    /// verdad: un contrato roto pasa a ser un fallo del productor.
    /// </summary>
    Strict,
}

/// <summary>Valida el <c>data</c> de un evento contra el esquema que declara.</summary>
/// <remarks>
/// La implementación vive en el paquete <c>Flux.Validation</c>. Esta interfaz está aquí
/// para que <see cref="ConnectOptions"/> pueda referirla sin que el paquete base dependa de
/// ninguna librería de JSON Schema.
/// </remarks>
public interface IEventValidator
{
    /// <summary>Comprueba el payload según el modo configurado.</summary>
    /// <param name="evento">El evento ya construido (al publicar) o ya parseado (al consumir).</param>
    /// <param name="subject">Subject del evento, para el mensaje de error.</param>
    /// <exception cref="SchemaValidationException">
    /// En <see cref="ValidationMode.Strict"/>, si el payload no cumple.
    /// </exception>
    /// <exception cref="SchemaNotFoundException">
    /// En <see cref="ValidationMode.Strict"/>, si el bundle no trae su esquema.
    /// </exception>
    void Check(FluxEvent evento, string subject);
}

/// <summary>Configuración de la validación L3.</summary>
/// <remarks>
/// El <see cref="Validator"/> se construye aparte porque el evaluador de esquemas vive en
/// otro paquete. La forma corta, con la extensión que ofrece <c>Flux.Validation</c>:
/// <code>
/// var options = new ConnectOptions
/// {
///     // …
///     Validation = new ValidationOptions
///     {
///         Mode      = ValidationMode.Strict,
///         Bundle    = SchemaBundle.FromFile("schemas/bundle.json"),
///         OnConsume = true,
///     }.WithSchemaValidator(),
/// };
/// </code>
/// </remarks>
public sealed record ValidationOptions
{
    /// <summary>Default <see cref="ValidationMode.Off"/>, es decir L2.</summary>
    public ValidationMode Mode { get; init; } = ValidationMode.Off;

    /// <summary>
    /// El bundle generado por <c>scripts/bundle-schemas.mjs</c>.
    /// </summary>
    /// <remarks>
    /// Se pasa como DATO: el SDK no resuelve el <c>dataschema</c> por HTTP ni con el modo
    /// estricto puesto — 00-protocol.md §5. Además de alimentar la validación, resuelve el
    /// <c>dataschema</c> exacto de cada subject al publicar.
    /// </remarks>
    public SchemaBundle? Bundle { get; init; }

    /// <summary>
    /// Validar también al CONSUMIR. Default <see langword="false"/>.
    /// </summary>
    /// <remarks>
    /// Un fallo se clasifica PERMANENT: el evento es sintácticamente correcto —ha llegado a
    /// parsearse, así que no es POISON— pero incumple su contrato, y reintentarlo dará
    /// exactamente el mismo resultado — 04-errors.md §1.2.
    /// </remarks>
    public bool OnConsume { get; init; }

    /// <summary>
    /// El evaluador de esquemas. Lo construye <c>Flux.Validation</c>.
    /// </summary>
    /// <remarks>
    /// <see cref="FluxBus.ConnectAsync"/> exige que no sea <see langword="null"/> cuando
    /// <see cref="Mode"/> no es <see cref="ValidationMode.Off"/>, y falla en el arranque
    /// diciendo qué paquete falta. Descubrir en la primera publicación que la validación
    /// que creías tener encendida no existía es peor que no tenerla: se cree que sí.
    /// </remarks>
    public IEventValidator? Validator { get; init; }

    /// <summary>
    /// Dónde van los avisos de <see cref="ValidationMode.Warn"/>.
    /// </summary>
    /// <remarks>
    /// <see langword="null"/> usa el <see cref="ConnectOptions.Logger"/> del bus. Si tampoco
    /// hay, el modo <c>Warn</c> no avisaría de nada y sería indistinguible de <c>Off</c>,
    /// así que <c>Flux.Validation</c> exige uno de los dos al construir el validador.
    /// </remarks>
    public IFluxLogger? Logger { get; init; }
}

/// <summary>
/// El <c>data</c> de un evento no valida contra su esquema.
/// </summary>
/// <remarks>
/// Al PUBLICAR sale de <see cref="FluxBus.PublishAsync"/> y aborta la publicación: es lo que
/// convierte un contrato roto en un fallo del servicio que lo provocó, en vez de un misterio
/// que aparece la semana que viene en un consumidor de otro equipo y otro lenguaje.
/// <para>
/// Al CONSUMIR se clasifica <see cref="ErrorClass.Permanent"/>: el evento es sintácticamente
/// correcto pero incumple su contrato, y reintentarlo seis veces daría exactamente el mismo
/// resultado bloqueando la cola 51 minutos para nada.
/// </para>
/// <para>
/// <see cref="Errors"/> trae TODOS los fallos, no solo el primero. Reportar de uno en uno
/// convierte arreglar un payload con tres campos mal en tres despliegues — 00-protocol.md §5.
/// </para>
/// </remarks>
public sealed class SchemaValidationException : ClassifiedException
{
    /// <summary>Código estable para métricas y alertas — 08-observability.md §2.2.</summary>
    public const string CodeValue = "SCHEMA_VALIDATION_FAILED";

    /// <summary>Construye la excepción.</summary>
    /// <param name="subject">Subject del evento.</param>
    /// <param name="dataSchema">URI del esquema contra el que se validó.</param>
    /// <param name="errors">TODOS los fallos encontrados, en orden estable.</param>
    public SchemaValidationException(string subject, string dataSchema, IReadOnlyList<string> errors)
        : base(BuildMessage(subject, dataSchema, errors), CodeValue, innerException: null)
    {
        Subject = subject;
        DataSchema = dataSchema;
        Errors = errors;
    }

    /// <inheritdoc />
    public override ErrorClass FluxClass => ErrorClass.Permanent;

    /// <summary>Subject del evento que no validó.</summary>
    public string Subject { get; }

    /// <summary>URI del esquema contra el que se validó.</summary>
    public string DataSchema { get; }

    /// <summary>TODOS los fallos de validación, no solo el primero.</summary>
    public IReadOnlyList<string> Errors { get; }

    private static string BuildMessage(string subject, string dataSchema, IReadOnlyList<string> errors)
    {
        var detalle = errors is null || errors.Count == 0
            ? string.Empty
            : "\n  · " + string.Join("\n  · ", errors);
        return $"el payload de \"{subject}\" no cumple su esquema ({dataSchema}):{detalle}";
    }
}

/// <summary>
/// El bundle no contiene el esquema que el evento dice cumplir.
/// </summary>
/// <remarks>
/// En <see cref="ValidationMode.Strict"/> es un error y no un aviso, y esa severidad es
/// deliberada: si "no lo encuentro" pasara en silencio, un bundle que se quedó atrás
/// convertiría la validación L3 en un no-op — el servicio seguiría arrancando, el panel
/// seguiría verde y NADA se estaría validando. Un fallo silencioso de la validación es peor
/// que no validar, porque se cree que sí.
/// </remarks>
public sealed class SchemaNotFoundException : ClassifiedException
{
    /// <summary>Código estable para métricas y alertas — 08-observability.md §2.2.</summary>
    public const string CodeValue = "SCHEMA_NOT_FOUND";

    /// <summary>Construye la excepción.</summary>
    /// <param name="subject">Subject del evento.</param>
    /// <param name="dataSchema">La URI que el evento declara y que el bundle no conoce.</param>
    public SchemaNotFoundException(string subject, string dataSchema)
        : base(
            $"no hay esquema para \"{subject}\" ({dataSchema}) en el bundle. " +
            "Regenéralo con `node scripts/bundle-schemas.mjs`, o baja el modo de validación a Warn",
            CodeValue,
            innerException: null)
    {
        Subject = subject;
        DataSchema = dataSchema;
    }

    /// <inheritdoc />
    public override ErrorClass FluxClass => ErrorClass.Permanent;

    /// <summary>Subject del evento.</summary>
    public string Subject { get; }

    /// <summary>La URI que no está en el bundle.</summary>
    public string DataSchema { get; }
}
